### PR TITLE
 Auto kill detached validator

### LINKED ISSUE #268: Spawn process that automatically kills detached validator after a timer
When starting a detached validator, also spawn a process that runs in the background which:

- Runs a timer (say an hour?)
- After the timer has elapsed, runs `kill-validator`, then itself exits (kills itself)
- Timer process can be reset by any roosevelt app in dev mode by:
  - Start/restart of any roosevelt app
  - Roosevelt apps should set an interval to refresh the timer process every so often so the validator won't get killed while an app is running

The desired effect will be if a developer kills his or her roosevelt app but forgets to run `npm run kill-validator`, the companion process will eventually kill the validator for them.

### PR DESCRIPTION
**Pull request by siuster95**
fixes #268

### CODE REVIEW

File: lib/defaults/config.json [hunk 1]
```diff
@@ -68,5 +68,6 @@
     "images"
   ],
   "versionedPublic": false,
-  "alwaysHostPublic": false
+  "alwaysHostPublic": false,
+  "autoKillerTime": 3600000
```
Comment (Autre31415) [inline lib/defaults/config.json]: Add this param to htmlValidator param group

**Comment by kethinov** [comment]
merge conflict

File: lib/scripts/autoKillValidator.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          logger.log('app is still active, resetting timer')
+          clearTimeout(timeoutHolder)
+          startTimeout()
+        }
+      })
+  }, timeout)
+}
+
+let filePath = path.join(__dirname, 'PID.txt')
+fse.writeFileSync(filePath, `${process.pid}`)
```
Comment (kethinov) [inline lib/scripts/autoKillValidator.js]: we should refactor this to set an environment variable instead, e.g.

```javascript
process.env['ROOSEVELT_VALIDATOR_KILLER_PID'] = process.pid
```

File: lib/scripts/autoKillValidator.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+const spawn = require('child_process').spawn
+const request = require('supertest')
```
Comment (kethinov) [inline lib/scripts/autoKillValidator.js]: is supertest really the right tool for the job here?

File: lib/scripts/autoKillValidator.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        } else {
+          logger.log('app is still active, resetting timer')
+          clearTimeout(timeoutHolder)
+          startTimeout()
+        }
+      })
+  }, timeout)
+}
+
+let tempObj = tmp.fileSync({keep: true, name: 'PID.txt'})
```
Comment (kethinov) [inline lib/scripts/autoKillValidator.js]: give this a more specific name, e.g. roosevelt_validator_pid

File: lib/scripts/autoKillValidator.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+fse.writeFileSync(tempObj.name, process.pid)
+logger.log(`Starting the auto Validator Killer, going to kill the validator in ${timeout / 1000} seconds if the app is not in use anymore`)
```
Comment (kethinov) [inline lib/scripts/autoKillValidator.js]: we need to add a feature that enables verbose logging for roosevelt, then restrict this log only to verbose mode

there are possibly other logs we should restrict to verbose mode too (or some to add and similarly restrict)

**Comment by kethinov** [comment]
also merge conflict

**Comment by kethinov** [comment]
another merge conflict

File: lib/scripts/autoKillValidator.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+const http = require('http')
+const fse = require('fs-extra')
+const path = require('path')
+const tmp = require('tmp')
+const os = require('os')
+
+let appPort = process.argv[2]
+let timeout = process.argv[3]
+let verbose = {verboseLogs: process.argv[4] === 'true'}
+console.log(verbose)
```
Comment (Autre31415) [inline lib/scripts/autoKillValidator.js]: Make sure these console.logs are removed

File: lib/scripts/autoKillValidator.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        startTimeout()
+      })
+
+    // if we get an error, likely that the connection is close and is safe to try to close the validator
+    }).on('error', () => {
+      logger.verbose('cannot connect to app, killing the validator now')
+      const killLine = spawn('node', [path.join(`${__dirname}/killValidator.js`)], {stdio: 'pipe', shell: false, windowsHide: true})
+
+      killLine.stdout.on('data', (data) => {
+        logger.verbose(`k stdout: ${data}`)
```
Comment (Autre31415) [inline lib/scripts/autoKillValidator.js]: Are these logging events supposed to be here?

File: README.md [hunk 8]
```diff
@@ -282,13 +282,15 @@ App behavior parameters
...
+  - `verboseLogs`: *[Boolean]* When set to true, Roosevelt will not output logs made by some of its other processes
   - Default: *[Object]*
 
       ```json
       {
         "httpLogs": false,
         "rooseveltLogs": false,
-        "rooseveltWarnings": false
+        "rooseveltWarnings": false,
+        "verboseLogs": false
```
Comment (kethinov) [inline README.md]: flip this to true

**Pull request merged**